File: clojure/src/defshef/blackjack.clj
(ns defshef.blackjack
  (:require [clojure.string :refer [trim]]
            [expectations :refer [expect in]]))

;; Modelling the base data

(def suits [:H :C :D :S])
(def ranks [2 3 4 5 6 7 8 9 10 :J :Q :K :A])

(defn card
  "build a card representation"
  [rank suit]
  {:pre [(some #{rank} ranks)
         (some #{suit} suits)]}
  {:rank rank
   :suit suit})

(expect 2 (:rank (card 2 :H)))
(expect :A (:rank (card :A :H)))
(expect :H (:suit (card 2 :H)))
(expect AssertionError (card 11 :H))
(expect AssertionError (card 7 :T))

(defn cards
  "Helper function for building a sequence of cards

  Usage: (cards 2 :H, 2 :C, 3 :S)
  "
  [& ranks-and-suits]
  (->> ranks-and-suits
       (partition 2)
       (mapv #(apply card %))))

(expect [(card 2 :H) (card :K :S)] (cards 2 :H, :K :S))
(expect [(card 2 :H) (card :K :S) (card :A :H)] (cards 2 :H, :K :S, :A :H))

;; Calculate hand value

(defn- card-value
  "How much is a card worth?

  In this function aces are always hard"
  [{:keys [rank] :as card}]
  (condp #(%1 %2) rank
    number? rank
    #{:J :Q :K} 10
    #{:A} 11
    (throw (ex-info "Unknown rank" {:card card}))))

(expect 5 (card-value (card 5 :C)))
(expect 10 (card-value (card 10 :H)))
(expect 10 (card-value (card :K :S)))
(expect 10 (card-value (card :Q :C)))
(expect 10 (card-value (card :J :D)))
(expect 11 (card-value (card :A :S)))
(expect Throwable (card-value {:rank :x}))

(def ^:private sum (partial reduce +))
(defn- is-ace [card] (-> card :rank (= :A)))
(defn- harden-aces
  "Attempt to un-bust by using low value of aces"
  [total aces]
  (if (or (<= total 21) (zero? aces))
    [total aces]
    (recur (- total 10) (dec aces))))

(defn value
  "How much is a hand worth?"
  [hand]
  (let [total (sum (map card-value hand))
        pair (= 2 (count hand))
        aces (count (filter is-ace hand))
        [total aces-remaining] (harden-aces total aces)
        qualifier (cond (and (= 21 total) pair) :blackjack
                        (> total 21) :bust
                        (pos? aces-remaining) :soft
                        (pos? aces) :hard
                        :else nil)]
    [total qualifier]))

(expect [4 nil] (value (cards 2 :H, 2 :C)))
(expect [8 nil] (value (cards 2 :H, 2 :C, 2 :D, 2 :S)))
(expect [19 nil] (value (cards 2 :H, 2 :C, 7 :D, 8 :S)))
(expect [21 nil] (value (cards :K :H, 5 :C, 6 :S)))
(expect [23 :bust] (value (cards :K :H, :Q :C, 3 :S)))
(expect [23 :bust] (value (cards 9 :H, :A :C, 3 :S, :K :C)))
(expect [27 :bust] (value (cards :J :H, 8 :C, 9 :S)))
(expect [21 :blackjack] (value (cards :K :H, :A :C)))
(expect [21 :blackjack] (value (cards :A :H, 10 :S)))
(expect [16 :soft] (value (cards :A :S, 5 :H)))
(expect [13 :hard] (value (cards :A :S, 5 :H, 7 :H)))
(expect [12 :soft] (value (cards :A :S, :A :H)))
(expect [14 :soft] (value (cards :A :S, :A :H, :A :C, :A :D)))
(expect [12 :hard] (value (cards :A :S, :A :H, :K :D)))
(expect [14 :hard] (value (cards :A :S, :A :H, :A :C, :A :D, :Q :H)))
(expect [24 :bust] (value (cards :A :S, :A :H, :A :C, :A :D, :Q :H, :J :H)))
(expect [21 :hard] (value (cards 10 :H, :Q :C, :A :S)))

;; Shuffle and deal

(def fresh-deck (for [s suits n ranks] (card n s)))

(expect (card 2 :H) (in fresh-deck))
(expect (card 6 :S) (in fresh-deck))
(expect (card :Q :D) (in fresh-deck))
(expect (card :A :C) (in fresh-deck))

(defn deal
  "Start a new game"
  []
  (let [deck (shuffle fresh-deck)
        [a b c d & deck] deck]
    {:deck deck
     :dealer [b d]
     :player [a c]
     :stage :player}))

(expect {:deck (drop 4 fresh-deck)
         :dealer (cards 3 :H, 5 :H)
         :player (cards 2 :H, 4 :H)
         :stage :player}
        (with-redefs [shuffle identity] (deal)))

(let [stacked-deck (cards :A :H, 7 :S, :Q :H, 6 :D, :K :S, :J :C)]
  (expect {:deck (cards :K :S, :J :C)
           :dealer (cards 7 :S, 6 :D)
           :player (cards :A :H, :Q :H)
           :stage :player}
          (with-redefs [shuffle (constantly stacked-deck)] (deal))))

;; Hit / Stand & basic gameplay

(defn hit
  "Advance the game state by hitting"
  [{:keys [deck player stage] :as game}]
  {:pre [(= stage :player)]}
  (let [[card & deck] deck
        player (conj player card)
        [_ qualifier] (value player)
        stage (if (= qualifier :bust) :dealer stage)]
    (assoc game :deck deck :player player :stage stage)))

(expect AssertionError (hit {:stage :dealer}))
(expect AssertionError (hit {:stage :done}))

; hit that doesn't bust
(expect {:deck (cards :J :C)
         :dealer (cards 7 :S, 6 :D)
         :player (cards 3 :H, 4 :H, :K :S)
         :stage :player}
        (hit {:deck (cards :K :S, :J :C)
              :dealer (cards 7 :S, 6 :D)
              :player (cards 3 :H, 4 :H)
              :stage :player}))

; hit that goes bust
(expect {:deck (cards :J :C)
         :dealer (cards 7 :S, 6 :D)
         :player (cards 10 :H, 4 :H, :K :S)
         :stage :dealer}
        (hit {:deck (cards :K :S, :J :C)
              :dealer (cards 7 :S, 6 :D)
              :player (cards 10 :H, 4 :H)
              :stage :player}))

(defn stand
  "Advance the game state by standing"
  [{:keys [stage] :as game}]
  {:pre [(= stage :player)]}
  (assoc game :stage :dealer))

(expect AssertionError (stand {:stage :dealer}))
(expect AssertionError (stand {:stage :done}))
(expect :dealer (:stage (stand (deal))))

(defn play-dealer
  "Advance the game state by playing the dealer's hand"
  [{:keys [stage] :as game}]
  {:pre [(= stage :dealer)]}
  (loop [{:keys [deck dealer] :as game} (assoc game :stage :done)]
    (let [[total _] (value dealer)]
      (if (< total 17)
        (let [[card & deck] deck]
          (recur (assoc game :deck deck :dealer (conj dealer card))))
        game))))

(expect AssertionError (play-dealer {:stage :player}))
(expect AssertionError (play-dealer {:stage :done}))

; dealer stands on 17+
(expect {:deck (cards :J :C)
         :dealer (cards 7 :S, :K :D)
         :player (cards 3 :H, 4 :H, :K :S)
         :stage :done}
        (play-dealer {:deck (cards :J :C)
                      :dealer (cards 7 :S, :K :D)
                      :player (cards 3 :H, 4 :H, :K :S)
                      :stage :dealer}))

; dealer hits on < 17
(expect {:deck (cards :Q :D)
         :dealer (cards 7 :S, 3 :D, :J :C)
         :player (cards 3 :H, 4 :H, :K :S)
         :stage :done}
        (play-dealer {:deck (cards :J :C, :Q :D)
                      :dealer (cards 7 :S, 3 :D)
                      :player (cards 3 :H, 4 :H, :K :S)
                      :stage :dealer}))

; dealer keeps going until 17+
(expect {:deck (cards 7 :D)
         :dealer (cards 4 :S, 3 :D, 3 :C, 4 :D, 2 :S, :J :D)
         :player (cards 3 :H, 4 :H, :K :S)
         :stage :done}
        (play-dealer {:deck (cards 3 :C, 4 :D, 2 :S, :J :D, 7 :D)
                      :dealer (cards 4 :S, 3 :D)
                      :player (cards 3 :H, 4 :H, :K :S)
                      :stage :dealer}))

(defn who-won
  ":player, :dealer, :draw ?"
  [{:keys [player dealer stage] :as game}]
  {:pre [(= stage :done)]}
  (let [[p-total p-qualifier] (value player)
        [d-total d-qualifier] (value dealer)]
    (cond (= :blackjack p-qualifier d-qualifier) :draw
          (= :blackjack d-qualifier) :dealer
          (= :blackjack p-qualifier) :player
          (= :bust p-qualifier) :dealer
          (= :bust d-qualifier) :player
          (= p-total d-total) :draw
          (> d-total p-total) :dealer
          (> p-total d-total) :player)))

(expect AssertionError (who-won {:stage :player}))
(expect AssertionError (who-won {:stage :dealer}))

(expect :dealer ; not bust, dealer higher
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 8 :H, 10 :C)
                  :player (cards 3 :H, 4 :H, :K :S)}))

(expect :player ; not bust, player higher
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 7 :H, 10 :C)
                  :player (cards 5 :H, 4 :H, :K :S)}))

(expect :draw ; not bust, same score
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 7 :H, 10 :C)
                  :player (cards 3 :H, 4 :H, :K :S)}))

(expect :player ; dealer bust
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 6 :H, 10 :C, 10 :D)
                  :player (cards 3 :H, 4 :H, :K :S)}))

(expect :dealer ; player bust
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 7 :H, 10 :C)
                  :player (cards 4 :H, :K :S, :J :C)}))

(expect :dealer ; both bust
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 6 :H, 10 :C, :J :D)
                  :player (cards 4 :H, :K :S, :J :C)}))

(expect :player ; blackjack
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 6 :H, 10 :C, 5 :D)
                  :player (cards 10 :H, :A :S)}))

(expect :dealer ; dealer blackjack
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 10 :H, :A :S)
                  :player (cards 6 :H, 10 :C, 5 :D)}))

(expect :draw ; both blackjack
        (who-won {:stage :done, :deck (cards)
                  :dealer (cards 10 :H, :A :S)
                  :player (cards :J :D, :A :C)}))

;; Display the game

(def ^:private card-faces
  {2 "🂢" 3 "🂣" 4 "🂤" 5 "🂥" 6 "🂦" 7 "🂧" 8 "🂨" 9 "🂩" 10 "🂪"
   :J "🂫" :Q "🂭" :K "🂮" :A "🂡"})
(def ^:private suit-offset
  {:S 0 :H 16 :D 32 :C 48})
(defn- adjust-suit [[x y :as face] suit]
  (str x (char (+ (int y) (suit-offset suit)))))

(defn- render-card
  "Render a single card"
  ([] "🂠 ")
  ([{:keys [rank suit]}]
   (-> rank (card-faces) (adjust-suit suit) (str " "))))

(expect "🂠 " (render-card))
(expect "🂡 " (render-card (card :A :S)))
(expect "🃍 " (render-card (card :Q :D)))

(defn- render-obscured
  "Render a hand with the first card obscured"
  [hand]
  (apply str
         (render-card)
         (map render-card (drop 1 hand))))

(defn- render-hand
  "Render a whole hand"
  [hand]
  (let [[total qualifier] (value hand)]
    (str (apply str (map render-card hand))
         " "
         (case qualifier
           :blackjack "Blackjack!"
           :bust (str total " BUST")
           (:hard :soft) (str (name qualifier) " " total)
           total))))

(defn- render-winner
  "Render the win announce line"
  [game]
  (case (who-won game)
    :dealer "Dealer wins!"
    :player "Player wins!"
    :draw "Draw"))

(defn render
  "Print out the game state neatly"
  [{:keys [dealer player stage] :as game}]
  (let [render-dealer (if (= stage :player) render-obscured render-hand)]
    (str "Dealer: " (render-dealer dealer) "\n"
         "Player: " (render-hand player)
         (if (= stage :done)
           (str "\n" (render-winner game))))))

(expect
  (str "Dealer: 🂠 🃉 " "\n"
       "Player: 🂾 🂥  15")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 7 :S, 9 :D)
           :player (cards :K :H, 5 :S)
           :stage :player}))

(expect
  (str "Dealer: 🂠 🃛 " "\n"
       "Player: 🂱 🃕  soft 16")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 7 :S, :J :C)
           :player (cards :A :H, 5 :C)
           :stage :player}))

(expect
  (str "Dealer: 🂠 🃛 " "\n"
       "Player: 🂱 🂪 🃕  hard 16")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 7 :S, :J :C)
           :player (cards :A :H, 10 :S, 5 :C)
           :stage :player}))

(expect
  (str "Dealer: 🂠 🃛 " "\n"
       "Player: 🂱 🃊  Blackjack!")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 7 :S, :J :C)
           :player (cards :A :H, 10 :D)
           :stage :player}))

(expect
  (str "Dealer: 🂠 🃛 " "\n"
       "Player: 🂶 🃊 🃋  26 BUST")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 7 :S, :J :C)
           :player (cards 6 :H, 10 :D, :J :D)
           :stage :player}))

(expect
  (str "Dealer: 🂨 🃉  17" "\n"
       "Player: 🂾 🂥  15")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 8 :S, 9 :D)
           :player (cards :K :H, 5 :S)
           :stage :dealer}))

(expect
  (str "Dealer: 🂥 🃛 🂮  25 BUST" "\n"
       "Player: 🂱 🃕  soft 16")
  (render {:deck (cards 6 :D)
           :dealer (cards 5 :S, :J :C, :K :S)
           :player (cards :A :H, 5 :C)
           :stage :dealer}))

(expect
  (str "Dealer: 🂥 🃛 🂮  25 BUST" "\n"
       "Player: 🂱 🃕  soft 16" "\n"
       "Player wins!")
  (render {:deck (cards 6 :D)
           :dealer (cards 5 :S, :J :C, :K :S)
           :player (cards :A :H, 5 :C)
           :stage :done}))

(expect
  (str "Dealer: 🂨 🃉  17" "\n"
       "Player: 🂾 🂥  15" "\n"
       "Dealer wins!")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 8 :S, 9 :D)
           :player (cards :K :H, 5 :S)
           :stage :done}))

(expect
  (str "Dealer: 🂦 🃛 🂴  20" "\n"
       "Player: 🂹 🂱  soft 20" "\n"
       "Draw")
  (render {:deck (cards :K :S, 6 :D)
           :dealer (cards 6 :S, :J :C, 4 :H)
           :player (cards 9 :H, :A :H)
           :stage :done}))

;; Tying it all together

(defn- player-turn [{:keys [player stage]}]
  (let [[_ qualifier] (value player)]
    (cond (not= stage :player) nil
          (#{:blackjack :bust} qualifier) '#{stand}
          :else '#{hit stand})))

(defn- read-action [actions]
  (apply println "Enter one of: quit" (map name actions))
  (let [input (symbol (trim (read-line)))]
    (cond (= 'quit input) (throw (ex-info "quit" {}))
          (actions input) (resolve input)
          :else (recur actions))))

(defn play-unnecessarily-generic-game [game possible-actions finish]
  (loop [game game]
    ; Check what choice of actions there is
    (let [actions (possible-actions game)]

      (if (seq actions)
        ; offer and apply choices
        (do
          (println (render game))
          (recur ((read-action actions) game)))

        ; no choices? finish the game
        (println (render (finish game)))))))

(defn play-blackjack [game]
  (play-unnecessarily-generic-game game player-turn play-dealer))

(defn start-game []
  (play-blackjack (deal)))

File: haskell/src/Blackjack.hs
module Blackjack
    ( shuffledDeck
    , cards
    , value
    ) where

import Text.Regex.Posix ((=~))
import Text.Read (readMaybe)
import Data.List.Split (endBy)
import System.Random.Shuffle as Shuffle

data Suit = Hearts | Clubs | Diamonds | Spades deriving (Eq, Enum)
data Rank = Pip Int | Jack | Queen | King | Ace deriving (Eq)
data Card = Card Rank Suit deriving (Show, Eq)

type Hand = [Card]
type Deck = [Card]

data HandQualifier = None | Soft | Hard | Bust | Blackjack deriving (Show, Eq)
data HandValue = HandValue Int HandQualifier deriving (Show, Eq)

data Stage = Player | Dealer | Done deriving (Show, Eq)
data Game = Game { deck :: Deck
                 , player :: Hand
                 , dealer :: Hand
                 , stage :: Stage
                 } deriving (Show, Eq)
data Winner = PlayerWins | DealerWins | Draw deriving (Show, Eq)

instance Show Suit where
    show Hearts = "H"
    show Clubs = "C"
    show Diamonds = "D"
    show Spades = "S"

instance Show Rank where
    show (Pip n) = show n
    show Jack = "J"
    show Queen = "Q"
    show King = "K"
    show Ace = "A"

instance Enum Rank where
    fromEnum (Pip n) = n
    fromEnum Jack = 11
    fromEnum Queen = 12
    fromEnum King = 13
    fromEnum Ace = 14
    toEnum n
        | n <= 10 = Pip n
        | n == 11 = Jack
        | n == 12 = Queen
        | n == 13 = King
        | n == 14 = Ace

-- | Build a card
--
-- >>> card Ace Hearts
-- Card A H
--
-- >>> card (Pip 10) Diamonds
-- Card 10 D
--
-- >>> card (Pip 20) Clubs
-- *** Exception: ...
--
-- >>> card (Pip 1) Clubs
-- *** Exception: ...
--
card :: Rank -> Suit -> Card
card (Pip n) s
    | n < 2 = error "Rank too low"
    | n > 10 = error "Rank too high"
card r s = Card r s

-- | Helper function for making a sequence of cards
--
-- I'm not entirely convinced that writing this was worth it.
-- Oh well.
--
-- >>> cards "AH KS"
-- [Card A H,Card K S]
cards :: String -> [Card]
cards = (map parseCard) . (endBy " ")

-- | Parse a card string into a Card
-- >>> parseCard "AH"
-- Card A H
-- >>> parseCard "6S"
-- Card 6 S
parseCard :: String -> Card
parseCard str = Card (parseRank r) (parseSuit s)
    where
        [[_, r, s]] = str =~ "^([2-9]|10|[JQKA])([HCDS])$" :: [[String]]

parseRank :: String -> Rank
parseRank "A" = Ace
parseRank "K" = King
parseRank "Q" = Queen
parseRank "J" = Jack
parseRank i = Pip (read i)

parseSuit :: String -> Suit
parseSuit "H" = Hearts
parseSuit "C" = Clubs
parseSuit "D" = Diamonds
parseSuit "S" = Spades

------------------------------------
--- Calculate hand value
------------------------------------

-- | Calculate hand value
--
-- >>> value (cards "2H 2C")
-- HandValue 4 None
-- >>> value (cards "2H 2C 2D 2S")
-- HandValue 8 None
-- >>> value (cards "2H 2C 7S 8D")
-- HandValue 19 None
-- >>> value (cards "KH 5C 6S")
-- HandValue 21 None
-- >>> value (cards "KH QC 3S")
-- HandValue 23 Bust
-- >>> value (cards "9H AC 3S KS")
-- HandValue 23 Bust
-- >>> value (cards "JH 8C 9S")
-- HandValue 27 Bust
-- >>> value (cards "AS KS")
-- HandValue 21 Blackjack
-- >>> value (cards "AS 5H")
-- HandValue 16 Soft
-- >>> value (cards "AS 5H 7C")
-- HandValue 13 Hard
-- >>> value (cards "AS AC")
-- HandValue 12 Soft
-- >>> value (cards "AS AC AH AD")
-- HandValue 14 Soft
-- >>> value (cards "AS AC KH")
-- HandValue 12 Hard
-- >>> value (cards "AS AC AH AD KC")
-- HandValue 14 Hard
-- >>> value (cards "AS AC AH AD KC JD")
-- HandValue 24 Bust
-- >>> value (cards "10H QC AS")
-- HandValue 21 Hard
value :: Hand -> HandValue
value hand = HandValue total qualifier
    where
        initialTotal = sum $ map cardValue hand
        ace = any isAce hand
        soft = ace && initialTotal <= 11
        total
            | soft = initialTotal + 10
            | otherwise = initialTotal
        pair = length hand == 2
        qualifier
            | total > 21 = Bust
            | total == 21 && pair = Blackjack
            | soft = Soft
            | ace = Hard
            | otherwise = None

-- | Calculate card value
cardValue :: Card -> Int
cardValue (Card (Pip n) _) = n
cardValue (Card rank _)
    | rank == Ace = 1
    | otherwise = 10

isAce :: Card -> Bool
isAce (Card Ace _) = True
isAce _ = False

------------------------------------
--- Shuffle & deal
------------------------------------

-- | The whole deck
freshDeck :: Deck
freshDeck = [Card r s | s <- [(Hearts)..Spades], r <- [Pip 2..Ace]]

-- | Make a new shuffled deck
shuffledDeck :: IO Deck
shuffledDeck = Shuffle.shuffleM freshDeck

-- | Start a game
-- >>> deal $ cards "AS 4D QH 6C 7H 3C"
-- Game {deck = [Card 7 H,Card 3 C], player = [Card A S,Card Q H], dealer = [Card 4 D,Card 6 C], stage = Player}
deal :: Deck -> Game
deal (a:b:c:d:leftover) = Game { deck=leftover
                               , player=[a, c]
                               , dealer=[b, d]
                               , stage=Player }


------------------------------------
--- Hit / Stand & basic gameplay
------------------------------------

-- | Hit
--
-- Takes another card:
-- >>> hit Game {deck = cards "AC", player = cards "6C 10D", dealer = cards "7S KC", stage = Player}
-- Game {deck = [], player = [Card 6 C,Card 10 D,Card A C], dealer = [Card 7 S,Card K C], stage = Player}
--
-- Moves to dealer if bust:
-- >>> hit Game {deck = cards "10C", player = cards "6C 10D", dealer = cards "7S KC", stage = Player}
-- Game {deck = [], player = [Card 6 C,Card 10 D,Card 10 C], dealer = [Card 7 S,Card K C], stage = Dealer}
hit :: Game -> Game
hit game = game { deck=remaining, player=player', stage=stage' }
    where
        (card:remaining) = deck game
        player' = player game ++ [card]
        HandValue _ qualifier = value player'
        stage'
            | qualifier == Bust = Dealer
            | otherwise = stage game

-- | Stand
--
-- Stops taking cards
-- >>> stand Game { deck = cards "AC", player = cards "8C 10D", dealer = cards "7S KC", stage = Player}
-- Game {deck = [Card A C], player = [Card 8 C,Card 10 D], dealer = [Card 7 S,Card K C], stage = Dealer}
stand :: Game -> Game
stand game = game { stage=Dealer }

-- | Play dealer
--
-- Dealer stands on 17
-- >>> playDealer Game {deck = cards "AC", player = cards "8C 10D", dealer = cards "7S KC", stage = Dealer}
-- Game {deck = [Card A C], player = [Card 8 C,Card 10 D], dealer = [Card 7 S,Card K C], stage = Done}
--
-- Dealer plays on under 17
-- >>> playDealer Game {deck = cards "6C", player = cards "8C 10D", dealer = cards "5S KC", stage = Dealer}
-- Game {deck = [], player = [Card 8 C,Card 10 D], dealer = [Card 5 S,Card K C,Card 6 C], stage = Done}
--
-- Dealer keeps playing until over 17
-- >>> playDealer Game {deck = cards "3C 4D 2S JD 7D", player = cards "8C 10D", dealer = cards "4S 3D", stage = Dealer}
-- Game {deck = [Card 7 D], player = [Card 8 C,Card 10 D], dealer = [Card 4 S,Card 3 D,Card 3 C,Card 4 D,Card 2 S,Card J D], stage = Done}
playDealer :: Game -> Game
playDealer game
    | total < 17 = playDealer $ game {deck=deck', dealer=dealer', stage=Dealer}
    | otherwise = game {deck=deck game, dealer=dealer game, stage=Done}
    where
        HandValue total _ = value $ dealer game
        (card:deck') = deck game
        dealer' = dealer game ++ [card]

-- | Who won?
--
-- not bust, dealer higher
-- >>> whoWon Game {deck = [], player = cards "8C 10D", dealer = cards "JS KC", stage = Done}
-- DealerWins
--
-- not bust, player higher
-- >>> whoWon Game {deck = [], player = cards "10C 10D", dealer = cards "9S KC", stage = Done}
-- PlayerWins
--
-- not bust, same score
-- >>> whoWon Game {deck = [], player = cards "10C 9D", dealer = cards "9S KC", stage = Done}
-- Draw
--
-- dealer bust
-- >>> whoWon Game {deck = [], player = cards "10C 9D", dealer = cards "6S KC JD", stage = Done}
-- PlayerWins
--
-- player bust
-- >>> whoWon Game {deck = [], player = cards "10C 5D 8C", dealer = cards "7S KC", stage = Done}
-- DealerWins
--
-- both bust
-- >>> whoWon Game {deck = [], player = cards "10C 5D 8C", dealer = cards "7S KC", stage = Done}
-- DealerWins
--
-- Blackjack
-- >>> whoWon Game {deck = [], player = cards "10C AC", dealer = cards "6S 5D KC", stage = Done}
-- PlayerWins
--
-- Dealer Blackjack
-- >>> whoWon Game {deck = [], player = cards "6S 5D KC", dealer = cards "10C AC", stage = Done}
-- DealerWins
--
-- Both Blackjack
-- >>> whoWon Game {deck = [], player = cards "AS KC", dealer = cards "10C AC", stage = Done}
-- Draw
whoWon :: Game -> Winner
whoWon game
    | pQualifier == Bust = DealerWins
    | dQualifier == Bust = PlayerWins
    | pQualifier == Blackjack && dQualifier == Blackjack = Draw
    | pQualifier == Blackjack = PlayerWins
    | dQualifier == Blackjack = DealerWins
    | pTotal > dTotal = PlayerWins
    | dTotal > pTotal = DealerWins
    | pTotal == dTotal = Draw
    where
        HandValue pTotal pQualifier = value $ player game
        HandValue dTotal dQualifier = value $ dealer game

------------------------------------
--- Display the game
------------------------------------


-- | Render the game to string
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 5D", player = cards "9D 8H", stage = Player }
-- Dealer: XX 5D
-- Player: 9D 8H (17)
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 6S", player = cards "AD 8H", stage = Player }
-- Dealer: XX 6S
-- Player: AD 8H (soft 19)
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 6S", player = cards "AD 8H 7D", stage = Player }
-- Dealer: XX 6S
-- Player: AD 8H 7D (hard 16)
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 6S", player = cards "AD 10H", stage = Player }
-- Dealer: XX 6S
-- Player: AD 10H (blackjack!)
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 6S", player = cards "6D 10H KC", stage = Player }
-- Dealer: XX 6S
-- Player: 6D 10H KC (26 BUST)
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 7S", player = cards "6D 10H KC", stage = Done }
-- Dealer: KH 7S (17)
-- Player: 6D 10H KC (26 BUST)
-- Dealer wins
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "KH 6S 8H", player = cards "6D 10H", stage = Done }
-- Dealer: KH 6S 8H (24 BUST)
-- Player: 6D 10H (16)
-- Player wins
--
-- >>> putStrLn $ render Game { deck = cards "AC", dealer = cards "9H AH", player = cards "6D 4C 10H", stage = Done }
-- Dealer: 9H AH (soft 20)
-- Player: 6D 4C 10H (20)
-- Draw
render :: Game -> String
render game =
    "Dealer: " ++ (renderDealer $ dealer game) ++ "\n" ++
    "Player: " ++ (renderHand $ player game) ++
    winLine
    where
        winLine = case stage game of
            Player -> ""
            _ -> "\n" ++ winner
        winner = case whoWon game of
            PlayerWins -> "Player wins"
            DealerWins -> "Dealer wins"
            _ -> "Draw"
        renderDealer = case stage game of
            Player -> renderObscured
            _ -> renderHand

renderHand :: Hand -> String
renderHand hand = cards ++ " (" ++ score ++ ")"
    where
        cards = unwords $ map renderCard hand
        HandValue total qualifier = value hand
        score = case qualifier of
            Blackjack -> "blackjack!"
            Bust -> show total ++ " BUST"
            Soft -> "soft " ++ show total
            Hard -> "hard " ++ show total
            _ -> show total

renderObscured :: Hand -> String
renderObscured (_:hand) = "XX " ++ (unwords $ map renderCard hand)

renderCard :: Card -> String
renderCard (Card r s) = show r ++ show s

------------------------------------
--- Tying it all together
------------------------------------

data Action = Hit | Stand deriving (Show, Read, Eq)
actualAction :: Action -> (Game -> Game)
actualAction Hit = hit
actualAction Stand = stand

playGame :: IO ()
playGame = do
    deck <- shuffledDeck
    gameLoop $ deal deck

gameLoop :: Game -> IO ()
gameLoop game = do
    putStrLn $ render game
    let actions = possibleActions game
    if null actions then do
        putStrLn $ render $ playDealer game
    else do
        actionName <- acceptAction actions
        let action = actualAction actionName
        gameLoop $ action game

acceptAction :: [Action] -> IO Action
acceptAction actions = do
    putStrLn $ "Choose one of " ++
        (unwords $ map show actions)
    actionStr <- getLine
    case readMaybe actionStr of
        Just action -> return action
        Nothing -> acceptAction actions

possibleActions :: Game -> [Action]
possibleActions game = case stage game of
    Player -> case value $ player game of
        HandValue _ Bust -> [Stand]
        HandValue _ Blackjack -> [Stand]
        HandValue _ _ -> [Hit, Stand]
    _ -> []
